google/cloud/forseti/scanner/audit/groups_settings_rules_engine.py

```diff
@@ -0,0 +1,471 @@
+    """Rule properties from the rule definition file, also finds violations."""
+
+    def __init__(self, rule_name, rule_index, rule):
+        """Initialize.
+
+        Args:
+            rule_name (str): Name of the loaded rule.
+            rule_index (int): The index of the rule from the rule definitions.
+            rule (dict): The rule definition from the file.
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.rule = rule
+
+    def rule_requirements(self):
+        """Used to create violation reason.
+
+       Returns:
+           str of property:value couples specified in rule,
+           joined by AND.
+        """
+        rule_list = []
+        for setting, value in self.rule['settings'].iteritems():
+            rule_list.append('{}:{}'.format(setting, value))
+        return ' AND '.join(rule_list)
+
+    def find_blacklist_violation(self, settings):
+        """Finds violations in case that rule is blacklist.
+        Args:
+            settings (GroupsSettings):
+        Returns:
+            violation_reason (str): Statement of what the broken rule required,
+                or empty string in case that rule is not violated.
+        """
+        violation_reason = ''
+        if not self.rule['settings']:
+            return violation_reason
+
+        violates_every_setting_in_rule = True
+        for setting, value in self.rule['settings'].iteritems():
+            if getattr(settings, setting) != value:
+                violates_every_setting_in_rule = False
+        if violates_every_setting_in_rule:
+            violation_reason = ("rule specified ({}) together is not allowed"
+                                .format(self.rule_requirements()))
+
+        return violation_reason
+
+    def find_whitelist_violation(self, settings):
+        """Finds violations in case that rule is whitelist.
+        Args:
+            settings (GroupsSettings):
+        Returns:
+            violation_reason (str): Statement of what the broken rule required,
+                or empty string in case that rule is not violated.
+        """
+        violation_reason = ''
+        a_setting_doesnt_match = False
+        for setting, value in self.rule['settings'].iteritems():
+            if getattr(settings, setting) != value:
+                a_setting_doesnt_match = True
+        if a_setting_doesnt_match:
+            violation_reason = ("rule specified ({}) is required"
+                                .format(self.rule_requirements()))
+
+        return violation_reason
+
+    def find_violations(self, settings):
+        """Find GroupsSettings violations.
+
+        Args:
+            settings (GroupsSettings): The resource to check for violations.
+
+        Returns:
+            list: Returns a list of RuleViolation named tuples.
+        """
+        violations = []
+        if self.rule['mode'] == BLACKLIST:
+            violation_reason = self.find_blacklist_violation(settings)
```

Comment:
Break down the code to check if violation is returned or not and then form the violation reason.